kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java

```diff
@@ -894,231 +893,84 @@ private void completeSendOperationForThrottling(long msgSize) {
-            completeSendOperationForThrottling(byteBuf.readableBytes());
-            encodeResult.recycle();
-            if (e == null) {
-                if (batch.isTransactional()) {
-                    getTransactionCoordinator().addActivePidOffset(TopicName.get(partitionName), batch.producerId(),
-                                                                    offset);
-                }
-                requestStats.getMessagePublishStats().registerSuccessfulEvent(
-                        MathUtils.elapsedNanos(beforePublish), TimeUnit.NANOSECONDS);
-                offsetConsumer.accept(offset);
-            } else {
-                log.error("publishMessages for topic partition: {} failed when write.", partitionName, e);
-                requestStats.getMessagePublishStats().registerFailedEvent(
-                        MathUtils.elapsedNanos(beforePublish), TimeUnit.NANOSECONDS);
-                errorsConsumer.accept(Errors.KAFKA_STORAGE_ERROR);
-            }
-        });
-    }
-
     @Override
     protected void handleProduceRequest(KafkaHeaderAndRequest produceHar,
                                         CompletableFuture<AbstractResponse> resultFuture) {
         checkArgument(produceHar.getRequest() instanceof ProduceRequest);
         ProduceRequest produceRequest = (ProduceRequest) produceHar.getRequest();
 
         final int numPartitions = produceRequest.partitionRecordsOrFail().size();
-
-        final Map<TopicPartition, PartitionResponse> responseMap = new ConcurrentHashMap<>();
-        // delay produce
-        final AtomicInteger topicPartitionNum = new AtomicInteger(produceRequest.partitionRecordsOrFail().size());
+        if (numPartitions == 0) {
+            resultFuture.complete(new ProduceResponse(Collections.emptyMap()));
+            return;
+        }
+        final Map<TopicPartition, PartitionResponse> unauthorizedTopicResponsesMap = new ConcurrentHashMap<>();
+        final Map<TopicPartition, MemoryRecords> authorizedRequestInfo = new ConcurrentHashMap<>();
         int timeoutMs = produceRequest.timeout();
-        Runnable complete = () -> {
-            topicPartitionNum.set(0);
-            if (resultFuture.isDone()) {
-                // It may be triggered again in DelayedProduceAndFetch
-                return;
-            }
-            // add the topicPartition with timeout error if it's not existed in responseMap
-            produceRequest.partitionRecordsOrFail().keySet().forEach(topicPartition -> {
-                if (!responseMap.containsKey(topicPartition)) {
-                    responseMap.put(topicPartition, new PartitionResponse(Errors.REQUEST_TIMED_OUT));
+        String namespacePrefix = currentNamespacePrefix();
+        final AtomicInteger unfinishedAuthorizationCount = new AtomicInteger(numPartitions);
+        Consumer<Runnable> completeOne = (action) -> {
+            // When complete one authorization or failed, will do the action first.
+            action.run();
```

Comment:
I think the `action` parameter here is not necessary. Instead of calling

```java
completeOne.accept(() -> someMap.put(someKey, someValue));
```

, I think the following code is more clear.

```java
someMap.put(someKey, someValue)
completeOne.run();
```

Reply:
You're right. Addressed.